### PR TITLE
Initializing a VilBERT model from a pre-trained transformer

### PR DESCRIPTION
**Pull request by matt-gardner**
Still a bunch of work left to do here:
- [x] I need to make sure I can load a range of different models and get the weight sizes to work out correctly (I may need to have vision sizes depend on text sizes, e.g.)
- [x] I need to actually initialize weights for the various components.  I'm not sure yet how hard this will be; hopefully it won't end up being a huge hack.  I might need to make a bunch of assumptions about the structure of the loaded model.

### CODE REVIEW

**Comment by matt-gardner** [comment]
Using albert as the base model passes tests.  My current hypothesis for zero gradients with BERT is that after a random initialization basically all activations get zeroed out by relus, so the model can't compute gradients.  I'm going to try getting the weights to load correctly from the pre-trained model, then see if that zero gradients error goes away.

**Comment by matt-gardner** [comment]
Initializing the weights from the pretrained model instead of default initializations worked, and the `"bert-base-uncased"` test now passes.  This is ready for review.

**Comment by matt-gardner** [comment]
Test error in CI seems unrelated; not sure if it's flaky or not.

**Comment by matt-gardner** [comment]
The test that's failing here passes in my local environment (and it's totally unrelated to this PR).  Maybe there's a new version of some dependency that's making the test fail...?

**Comment by dirkgr** [comment]
I'm hoping that test will succeed with an update from `master`.

File: allennlp/data/dataset_readers/nlvr2_reader.py [hunk 1]
```diff
@@ -106,17 +106,21 @@ def __init__(
         text_path_prefix: str,
         visual_path_prefix: str,
         topk_images: int = -1,
+        tokenizer: Tokenizer = None,
+        token_indexers: Dict[str, TokenIndexer] = None,
```
Comment (dirkgr) [inline allennlp/data/dataset_readers/nlvr2_reader.py:110]: If the only correct choices here are of the `PretrainedTransformer*` variety, and they have to match, then I think this should only take `model_name: str = "bert-base-uncased"` parameter.

File: allennlp/models/vilbert.py [hunk 2]
```diff
@@ -714,6 +789,95 @@ def __init__(
...
+        # to do if it fails.  We should probably pull this out into a central "transformers_util"
+        # module, or something, and just have a method that pulls out an initialized embedding layer
+        # from a huggingface model.
+        text_embeddings = deepcopy(transformer.embeddings)
+        if hasattr(transformer.config, "embedding_size"):
+            config = transformer.config
+
+            from transformers.modeling_albert import AlbertModel
+
+            if isinstance(transformer, AlbertModel):
```
Comment (dirkgr) [inline allennlp/models/vilbert.py:827]: The comment above says this is just for BERT, but here you have Albert?

File: allennlp/models/vilbert.py [hunk 3]
```diff
@@ -714,6 +789,95 @@ def __init__(
...
+            # because Sequential doesn't accept *inputs, only a single argument.
+
+            class EmbeddingsShim(torch.nn.Module):
+                def __init__(self, embeddings: torch.nn.Module, linear_transform: torch.nn.Module):
+                    super().__init__()
+                    self.linear_transform = linear_transform
+                    self.embeddings = embeddings
+
+                def forward(self, *inputs, **kwargs):
+                    return self.linear_transform(self.embeddings(*inputs, **kwargs))
```
Comment (dirkgr) [inline allennlp/models/vilbert.py:846]: These embeddings are just a linear transform on top of existing pre-trained embeddings? What's the point of that?

Are the embeddings shared in some other place, and that's why? If these embedding weights are supposed to be tied to some other weights somewhere, I'm pretty sure that the `deepcopy` above breaks that.

**Comment by dirkgr** [comment]
> I'm hoping that test will succeed with an update from `master`.

Looks like that worked!

File: allennlp/data/dataset_readers/nlvr2_reader.py (hunk 1 above)
Comment (matt-gardner) [reply allennlp/data/dataset_readers/nlvr2_reader.py:110]: Yeah, I thought about that, though I think there's only one place that we actually require this, and it's in logic that only exists because of a small research project that Sanjay did.  I'm not sure we need to keep that logic in this version of the class.

To be more concrete: a `PretrainedTransformerTokenizer` is only required if you use `mask_prepositions_verbs`.  Instead of restricting the types to only be `PretrainedTransformer*` objects here, I'd instead either:
1. Remove the `{drop,mask}_prepositions_verbs` functionality from this reference implementation as not needed
2. Raise an error if you don't have the right tokenizer when using `mask_prepositions_verbs`
3. Come up with a workaround for how to mask prepositions and verbs when you're not using a pretrained transformer (though this probably doesn't make much sense).

I put those in my order of preference.  I'm not feeling too well today, so I'm not going to be working on this, but I'll do option 1 the next time I write code, unless anyone has objections.

File: allennlp/models/vilbert.py (hunk 2 above)
Comment (matt-gardner) [reply allennlp/models/vilbert.py:827]: Thanks, yep, I need to update that comment.

File: allennlp/models/vilbert.py (hunk 3 above)
Comment (matt-gardner) [reply allennlp/models/vilbert.py:846]: The pipeline in a transformer is essentially:

```
hidden_states = self.embeddings(inputs)
for layer in self.layers:
    hidden_states = layer(hidden_states)
```

For this logic to work on its own, `embedding_dim` must equal `hidden_dim`, and this is what BERT (and most other transformers) do.  Albert decides to have a much smaller `embedding_dim`, with a linear transform after the embedding to get the hidden states up to the right size:

```
hidden_states = self.linear_transform(self.embeddings(inputs))
for layer in self.layers:
    hidden_states = layer(hidden_states)
```

This reduces the number of parameters in your model: embedding matrix goes from `(vocab_size, 768)` to `(vocab_size, 128)`, while you add a `(128, 768)` linear transform matrix.

There is nothing that's supposed to be tied anywhere; I used `deepcopy` here to get a new copy of these existing weight matrices (both the embeddings and the linear transform), to initialize the weights in our model correctly.  Duplicating the modules themselves is much easier than aligning all of the underlying `Parameters` and calling `.copy_()`.

Comment (matt-gardner) [reply allennlp/models/vilbert.py:846]: Hmm, it's possible there is an activation in between the embeddings and the transform in albert; I should double check that.

Comment (dirkgr) [reply allennlp/models/vilbert.py:846]: Makes sense :+1:

File: allennlp/data/dataset_readers/nlvr2_reader.py (hunk 1 above)
Comment (dirkgr) [reply allennlp/data/dataset_readers/nlvr2_reader.py:110]: I agree that this functionality might not be needed in this reference implementation, but that's orthogonal to how the tokenizer and indexer are specified, isn't it?

Comment (matt-gardner) [reply allennlp/data/dataset_readers/nlvr2_reader.py:110]: Yes, but it's that functionality that is the only thing placing constraints on this.

Oh, unless you're saying that the model requires the use of a pretrained transformer?  Well, vilbert does, but you could use this dataset reader with other models that don't make that assumption.  There's nothing in the reader itself that requires that use, so I was keeping as much flexibility in the types as possible.

File: allennlp/models/vilbert.py (hunk 3 above)
Comment (matt-gardner) [reply allennlp/models/vilbert.py:846]: I just checked, and there is no activation in between the embedding and the hidden layer in albert, so this logic is correct.

**Comment by matt-gardner** [comment]
@dirkgr, this should be ready for another look.

File: allennlp/models/vilbert.py [hunk 4]
```diff
@@ -814,11 +814,17 @@ def from_huggingface_model_name(
...
+        # not sure it works for all of them, or what to do if it fails.
+        # We should probably pull everything up until the instantiation of the image feature
+        # embedding out into a central "transformers_util" module, or something, and just have a
+        # method that pulls an initialized embedding layer out of a huggingface model.  One place
+        # for this somewhat hacky code to live, instead of having to duplicate it in various models.
         text_embeddings = deepcopy(transformer.embeddings)
+
+        # Albert (and maybe others?) has this "embedding_size", that's different from "hidden_size".
+        # To get them to the same dimensionality, it uses a linear transform after the embedding
+        # layer, which we need to pull out and copy here.
```
Comment (dirkgr) [inline allennlp/models/vilbert.py:827]: These seem like excellent concerns to handle in @AkshitaB's transformer toolkit? We could have a function that says "give me the embedding layer from "roberta-large" or "albert-v2", and it returns a `torch.nn.Module` that does the right thing?

Comment (matt-gardner) [reply allennlp/models/vilbert.py:827]: Yes, agreed, that was the point of the comment starting on line 819.
